cloud/ec2.go

```diff
@@ -656,7 +656,9 @@ func (m *ec2Manager) GetDNSName(h *host.Host) (string, error) {
 		if err != nil {
 			return "", errors.Wrapf(err, "failed to get spot request info for %s", h.Id)
 		}
-
+		if spotDetails.SpotInstanceRequests[0].InstanceId == nil {
+			return "", errors.WithStack(errors.New("spot instance does not yet have an instanceId"))
+		}
 		if *spotDetails.SpotInstanceRequests[0].InstanceId != "" {
```

Comment:
Just curious: What does mean `*` before expression? And why the line with `nil` check doesn't have it?

Reply:
`InstanceId` is a pointer to a string, not a string (https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotInstanceRequest). The AWS SDK for Go uses this idiom a lot. Typically in Go you expect that the zero value for a string will be the empty string, but the AWS SDK wants to distinguish between not setting a value at all and setting the value to the empty string, so they use pointers to values instead of just values. The bug here is that I was dereferencing a pointer to a string without first checking it is nil.

Reply:
`NullPointerException` - classics =)
I wish to have Kotlin-like experience for null-safety.